packages/karma-runner/src/karma-test-runner.ts

```diff
@@ -1,77 +1,69 @@
-import karma from 'karma';
-import { StrykerOptions, MutantCoverage } from '@stryker-mutator/api/core';
+import { StrykerOptions } from '@stryker-mutator/api/core';
 import { Logger, LoggerFactoryMethod } from '@stryker-mutator/api/logging';
 import { commonTokens, tokens } from '@stryker-mutator/api/plugin';
-import {
-  TestRunner,
-  TestResult,
-  DryRunStatus,
-  DryRunOptions,
-  MutantRunOptions,
-  DryRunResult,
-  MutantRunResult,
-  toMutantRunResult,
-} from '@stryker-mutator/api/test-runner';
+import { TestRunner, DryRunOptions, MutantRunOptions, DryRunResult, MutantRunResult, toMutantRunResult } from '@stryker-mutator/api/test-runner';
+import type { Config } from 'karma';
 
 import { StrykerKarmaSetup } from '../src-generated/karma-runner-options';
 
-// eslint-disable-next-line @typescript-eslint/no-require-imports
-import strykerKarmaConf = require('./starters/stryker-karma.conf');
+import { karma } from './karma-wrapper';
+import strykerKarmaConf from './starters/stryker-karma.conf';
 import { ProjectStarter } from './starters/project-starter';
 import { StrykerReporter } from './karma-plugins/stryker-reporter';
 import { KarmaRunnerOptionsWithStrykerOptions } from './karma-runner-options-with-stryker-options';
 import { TestHooksMiddleware } from './karma-plugins/test-hooks-middleware';
 
-export interface ConfigOptions extends karma.ConfigOptions {
-  detached?: boolean;
-}
-
 export class KarmaTestRunner implements TestRunner {
-  private currentTestResults: TestResult[] = [];
-  private currentErrorMessage: string | undefined;
-  private currentCoverageReport?: MutantCoverage;
   private readonly starter: ProjectStarter;
-  public port: number | undefined;
+  private exitPromise: Promise<number> | undefined;
+  private runConfig!: Config;
 
   public static inject = tokens(commonTokens.logger, commonTokens.getLogger, commonTokens.options);
   constructor(private readonly log: Logger, getLogger: LoggerFactoryMethod, options: StrykerOptions) {
     const setup = this.loadSetup(options);
     this.starter = new ProjectStarter(getLogger, setup);
     this.setGlobals(setup, getLogger);
-    this.cleanRun();
-    this.listenToServerStart();
-    this.listenToSpecComplete();
-    this.listenToCoverage();
-    this.listenToError();
   }
 
   public async init(): Promise<void> {
-    return new Promise((res, rej) => {
-      StrykerReporter.instance.once('browsers_ready', () => res());
-      this.starter.start().catch(rej);
+    const browsersReadyPromise = StrykerReporter.instance.whenBrowsersReady();
+    const { exitPromise } = await this.starter.start();
+    this.exitPromise = exitPromise;
```

Comment:
We should handle premature exits. It was done (in a dirty way) before, but it got lost in translation during the refactor. We should be able to add an integration test for this. For example, add a framework that crashes stuff 🤷‍♂️